internal/v4/auth.go

```diff
@@ -61,14 +72,61 @@ func (h *Handler) authenticate(w http.ResponseWriter, req *http.Request, id *cha
 	return httpbakery.NewDischargeRequiredError(m, verr)
 }
 
+func (h *Handler) authorizeEntity(id *charm.Reference, req *http.Request) error {
+	// TODO frankban: the first time a new charm is published, its
+	// corresponding base entity is not yet present in the database.
+	// For this reason, the check below must still allow specific users to
+	// proceed with the request, even in the case ACL cannot be retrieved
+	// from the base entity.
+	baseEntity, err := h.store.FindBaseEntity(id, "acls")
+	if err != nil {
+		if errgo.Cause(err) == params.ErrNotFound {
+			// Cannot get the ACL from the entity.
+			// TODO frankban: just assume read permissions for everyone and
+			// no write permissions for now. Let the endpoint deal with not
+			// found errors.
+			return h.authorizeWithPerms(req, []string{params.Everyone}, nil)
+		}
+		return errgo.Notef(err, "cannot retrieve entity %q for authorization", id)
+	}
+	// TODO frankban: replace with baseEntity.ACL.Write.
+	// For the time being only rely on basic HTTP auth.
+	return h.authorizeWithPerms(req, baseEntity.ACLs.Read, nil)
+}
+
+func (h *Handler) authorizeWithPerms(req *http.Request, read, write []string) error {
+	var acl []string
+	switch req.Method {
+	case "DELETE", "PATCH", "POST", "PUT":
+		acl = write
+	default:
+		acl = read
+	}
+	return h.authorize(req, acl)
+}
+
+const userNameAttr = "username"
+
+func (h *Handler) checkACLMembership(attrs map[string]string, acl []string) error {
+	user := attrs[userNameAttr]
+	if user == "" {
+		return errgo.New("no username declared")
+	}
+	for _, name := range acl {
+		if name == params.Everyone || name == user {
+			return nil
+		}
+	}
+	return errgo.Newf("access denied for user %q", user)
+}
+
 func (h *Handler) newMacaroon() (*macaroon.Macaroon, error) {
 	// TODO generate different caveats depending on the requested operation
 	// and whether there's a charm id or not.
 	// Mint an appropriate macaroon and send it back to the client.
 	return h.store.Bakery.NewMacaroon("", nil, []checkers.Caveat{{
-		Location: h.config.AuthLocation,
-		// TODO needs-declared user is-authenticated-user
-		Condition: "is-authenticated-user",
+		Location:  h.config.AuthLocation,
+		Condition: "need-declared " + userNameAttr + " is-authenticated-user",
```

Comment:
You should probably use http://godoc.org/gopkg.in/macaroon-bakery.v0/bakery/checkers#NeedDeclaredCaveat for this.